src/coreclr/nativeaot/System.Private.CoreLib/src/System/RuntimeExceptionHelpers.cs

```diff
@@ -235,71 +224,42 @@ internal static void FailFast(string message, Exception? exception, RhFailFastRe
-                        // Random 10 bytes to identify the record
-                        ((ReadOnlySpan<byte>)new byte[] { 0x1c, 0x73, 0xd0, 0x2d, 0xda, 0x6b, 0x4c, 0xef, 0xbf, 0xa1 }).CopyTo(cookieSpan);
-                        "NETRUNTIME"u8.CopyTo(cookieSpan.Slice(10));
-                        pExceptionRecord->Type = 1;
-                        pExceptionRecord->Data = (void*)exceptionStack.GetPinnableReference();
-                        pExceptionRecord->Length = length;
-                    }
+                    _crashInfo.WriteExceptionInfo(exception);
                 }
             }
 
-#if TARGET_WINDOWS
-            uint errorCode = 0x80004005; // E_FAIL
-            // To help enable testing to bucket the failures we choose one of the following as errorCode:
-            // * hashcode of EETypePtr if it is an unhandled managed exception
-            // * HRESULT, if available
-            // * RhFailFastReason, if it is one of the known reasons
-            if (exception != null)
-            {
-                if (reason == RhFailFastReason.UnhandledException)
-                    errorCode = (uint)(exception.GetEETypePtr().GetHashCode());
-                else if (exception.HResult != 0)
-                    errorCode = (uint)exception.HResult;
-            }
-            else if (reason != RhFailFastReason.Unknown)
+            _crashInfo.Close();
+
+            // Try to map the failure into a HRESULT that makes sense
+            int errorCode = exception != null ? exception.HResult : reason switch
             {
-                errorCode = (uint)reason + 0x1000; // Add something to avoid common low level exit codes
-            }
+                RhFailFastReason.EnvironmentFailFast => HResults.COR_E_APPLICATION,
+                RhFailFastReason.InternalError or
+                RhFailFastReason.ClassLibDidNotTranslateExceptionID => HResults.COR_E_EXECUTIONENGINE,
+                RhFailFastReason.UnhandledException or
+                RhFailFastReason.UnhandledExceptionFromPInvoke or
+                RhFailFastReason.UnhandledException_ExceptionDispatchNotAllowed or
+                RhFailFastReason.UnhandledException_CallerDidNotHandle => HResults.E_ACCESSDENIED,
+                _ => HResults.E_FAIL
+            };
 
-            Interop.Kernel32.RaiseFailFastException(errorCode, pExAddress, pExContext);
+#if TARGET_WINDOWS
+            Interop.Kernel32.RaiseFailFastException(errorCode, pExAddress, pExContext, _crashInfo.GetTriageBuffer(out int size), size);
 #else
             Interop.Sys.Abort();
 #endif
         }
 
-        private unsafe struct CrashDumpRecord
-        {
-            public const int CookieSize = 20;
-            public fixed byte Cookie[CookieSize];
-            public int Type;
-            public void* Data;
-            public int Length;
-        }
+        private static CrashInfo _crashInfo = new();
```

Comment:
```suggestion
        private static CrashInfo s_crashInfo = new();
```
Nit: Coding conventions.

Alternatively, it may be ok to just allocate the instance of CrashInfo on the GC heap instead of pre-allocating it. It is very small.

Reply:
Quick question for further work in this path - how should we reason about allocations in error paths, i.e. in cases where Stack OverFlow or OOM might have been a cause?

Reply:
I was thinking about this some more and for OOM (stack overflow doesn't go through this code I'm pretty sure), it may be better to statically allocate the CrashInfo object like I did originally.  Then there shouldn't be any GC allocations writing the JSON header and closing bracket. Writing the exception info is skipped because  minimalFailFast is true.  Watson would get some minimal info.

Reply:
Static constructors are executed lazily by default. It means that the `private static CrashInfo s_crashInfo = new();` is going to be executed lazily first time when you access `s_crashInfo`.

Can you turn `CrashInfo` into a struct? It would be allocated on the stack and the OOM problem would go away.